text/0047-random-assignment-of-availability-chunks.md

```diff
@@ -0,0 +1,212 @@
+It needs to be explicitly stated that after the runtime upgrade and governance enactment, validators that run older
+client versions that don't support this mapping will not be able to participate in parachain consensus.
+
+### Getting access to core_index
+
+Availability-recovery can currently be triggered by the following steps in the polkadot protocol:
+1. During the approval voting process.
+1. By other collators of the same parachain
+3. During disputes.
+
+The `core_index` refers to the index of the core that the candidate was occupying while it was pending availability
+(from backing to inclusion).
+Getting the right core index for a candidate can prove troublesome. Here's a breakdown of how different parts of the
+node implementation can get access to this data:
+
+1. The approval-voting process for a candidate begins after observing that the candidate was included. Therefore, the
+node has easy access to the block where the candidate got included (and also the core that it occupied).
+1. The `pov_recovery` task of the collators starts availability recovery in response to noticing a candidate getting
+backed, which enables easy access to the core index the candidate started occupying.
+1. Disputes may be initiated on a number of occasions:
+  
+    3.a. is initiated by the validator as a result of finding an invalid candidate while participating in the
+  approval-voting protocol. In this case, availability-recovery is not needed, since the validator already issued their
+  vote.
+    
+    3.b is initiated by the validator noticing dispute votes recorded on-chain. In this case, we can safely
+  assume that the backing event for that candidate has been recorded and kept in memory.
+  
+    3.c is initiated as a result of getting a dispute statement from another validator. It is possible that the dispute
+  is happening on a fork that was not yet imported by this validator, so the subsystem may not have seen this candidate
+  being backed.
+
+As a solution to 3.c, a new version for the disputes request-response networking protocol will be added.
+This message type will include the relay block hash where the candidate was included. This information will be used
+in order to query the runtime API and retrieve the core index that the candidate was occupying.
+
+The usage of the systematic data availability recovery feature will also be subject to all nodes using the V2 disputes
+networking protocol.
+
+#### Alternatives to using core_index
+
+As an alternative to core_index, the `ParaId` could be used. It has the advantage of being readily available in the
+`CandidateReceipt`, which would enable the dispute communication protocol to not change and would simplify the
+implementation.
+However, in the context of [CoreJam](https://github.com/polkadot-fellows/RFCs/pull/31), `ParaId`s will no longer exist
+(at least not in their current form).
```

Comment:
IMHO we shouldn't worry too much about CoreJam compatibility. IIUC there might be a drop-in replacement for `ParaId` (`AuthId`), so we should first explore an avenue with `ParaId` perhaps. 

The (bigger) problem with `ParaId` is that it's claimable by users, so in the theory one could create a collision attack where multiple paras have the same systematic chunks. In practice, I believe such an attack would be high cost and low benefit. And, perhaps, it could be mitigated by using a cryptographic hash of `ParaId` (or `AuthId` or `core_index`) in the future.

Reply:
We should use the core index here probably, not the para id.  I'd thought the core indexes could be spaced somewhat evenly, but with the actual sequence being random.

Reply:
I changed my view here above in https://github.com/polkadot-fellows/RFCs/pull/47#discussion_r1433047970  

It's probably more future proof if we do not require the core index here, so the map uses num_cores, num_validators, relay parent, and paraid.  We've other options but this looks pretty future proof.  CoreJam cannot really work without something like paraid.

Reply:
Using paraid was my first attempt, but it complicates things quite a bit.
As Andronik notes, it's claimable by a user. I don't think it's the biggest problem though, they can't get a different ParaId every day (can they?).

`paraid`'s biggest problem is that it doesn't form a strictly monontonically increasing sequence, so we'd have to do something else with it. My initial attempt was to seed a ChaCha8 RNG with it and generate one random number. Then this number was the offset into the chunk index vector.

But then if we complicate it so much, we may want to embed it into the runtime, so that future updates to the `rand` crate don't break us (which creates all sorts of problems that could be avoided).

We've had this paraid -> core_index back and forth for a while now, it'd be great if we could all (the ones interested in this RFC) hop on a call or somehow reach a conclusion. There are pros and cons to both, AFAICT.

I think this remains the only bit that prevents this RFC from moving on (anybody correct me if I'm wrong)